apis/v1alpha1/validation/validation.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"net"
+	"strings"
+
+	gatewayv1a1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateGateway validates Gateway objects.
+func ValidateGateway(gw *gatewayv1a1.Gateway) field.ErrorList {
+	return ValidateGatewaySpec(&gw.Spec, field.NewPath("spec"))
+}
+
+// ValidateGatewaySpec validates whether required fields of spec are set according to the
+// Gateway API specification.
+func ValidateGatewaySpec(spec *gatewayv1a1.GatewaySpec, path *field.Path) field.ErrorList {
+	// TODO [danehans]: Add additional validation of spec fields.
+	return validateGatewayListeners(spec.Listeners, path.Child("listeners"))
+}
+
+// validateGatewayListeners validates whether required fields of listeners are set according
+// to the Gateway API specification.
+func validateGatewayListeners(listeners []gatewayv1a1.Listener, path *field.Path) field.ErrorList {
+	// TODO [danehans]: Add additional validation of listener fields.
+	return ValidateListenerHostname(listeners, path)
+}
+
+// ValidateListenerHostname validates each listener hostname is not an IP address and is one
+// of the following:
+//  - A fully qualified domain name of a network host, as defined by RFC 3986.
+//  - A DNS subdomain as defined by RFC 1123.
+//  - A wildcard DNS subdomain as defined by RFC 1034 (section 4.3.3).
+func ValidateListenerHostname(listeners []gatewayv1a1.Listener, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	for i, h := range listeners {
```

Comment:
suggest early breaks to improve readability:

```Go
for i, h := range listeners {
        if h.Hostname == nil {
                continue
        }
        switch *h.Hostname {
        case  "": 
                // is this really valid?
        case "*":
                // valid wildcard
        default:
                // parse and check
        }
} 
```

Reply:
I refactored the conditional logic that I think makes it easier to follow.